src/main/java/htsjdk/samtools/CompressedIndexFileBuffer.java

```diff
@@ -0,0 +1,84 @@
+package htsjdk.samtools;
+
+import htsjdk.samtools.util.BlockCompressedInputStream;
+import htsjdk.samtools.util.RuntimeIOException;
+
+import java.io.File;
+import java.io.IOException;
+
+
+class CompressedIndexFileBuffer implements IndexFileBuffer {
+
+    private BlockCompressedInputStream mCompressedStream;
+
+    CompressedIndexFileBuffer(File file) {
+        try {
+            mCompressedStream = new BlockCompressedInputStream(file);
+        } catch (IOException ioe) {
+            throw(new RuntimeIOException("Construction error of CSI compressed stream: " + ioe));
+        }
+    }
+
+    @Override
+    public void readBytes(final byte[] bytes) {
+        try {
+            mCompressedStream.read(bytes);
+        } catch (IOException ioe) {
+            throw(new RuntimeIOException("Read error in CSI compressed stream: " + ioe));
+        }
+    }
+
+    @Override
+    public int readInteger() {
+        final byte[] intbuff = new byte[4];
```

Comment:
Allocating a new byte[] for each integer read is going to be super inefficient.   It would make sense to reuse the buffer.  This method suffers from the same problem with read not guaranteeing a definite number of bytes.